### DISCUSSION TITLE
Properly setting "GRIB2_ipdtmpl_index" and "GRIB2_ipdtmpl_val" in conf file

### DISCUSSION THREAD

**Discussion by michelleharrold** [dtcenter/METplus#1241] [Configuration] [answered]
Hi There,

I am struggling to properly set the "GRIB2_ipdtmpl_index" and "GRIB2_ipdtmpl_val" options for specifying the correct field in my GRIB2 output. I am trying to verify total cloud (TCDC) using prepbufr data (run through pb2nc).

When I first ran point-stat, I got this warning:

```
WARNING:
WARNING: MetGrib2DataFile::data_plane_array() -> Using the first of 2 exact matching records for "TCDC/L0":
WARNING: record 914 field 1: ipdtmpl[15] = 6, 1, 2, 0, 134, 0, 0, 1, 7, 200, 0, 0, 255, 0, 0
WARNING: record 915 field 1: ipdtmpl[29] = 6, 1, 2, 0, 134, 0, 0, 1, 6, 200, 0, 0, 255, 0, 0, 2019, 6, 15, 7, 0, 0, 1, 0, 0, 2, 1, 1, 255, 0
WARNING: Try setting the 0-based "GRIB2_ipdtmpl_index" and "GRIB2_ipdtmpl_val" config file options to select one.
WARNING:
```

While, I do, in fact, want the first of the 2 matching records, I decided to update the logic in the METplus conf file to make it specific enough to only find one field. Here is how I specified the FCST field:

```
FCST_VAR8_NAME = TCDC
FCST_VAR8_LEVELS = L0
FCST_VAR8_OPTIONS = GRIB_lvl_typ = 200; GRIB2_ipdtmpl_index = [ 14 ]; GRIB2_ipdtmpl_val = [ 0 ];
```

However, I am still getting the same warning -- so clearly I need to up my knowledge game with the GRIB2_ipdtmpl_index and GRIB2_ipdtmpl_val settings. I tried several numeric combinations for the settings, but they do not yield the desired result. I tried navigating through some of the [GRIB2 documentation](https://www.nco.ncep.noaa.gov/pmb/docs/grib2/grib2_doc/grib2_table4-0.shtml), but I still struggling.

Any help would be greatly appreciated!
Michelle Harrold

**Comment by j-opatz** [comment]
Hi Michelle,

And thank you for your question. It looks like there may be a bit of confusion on what the GRIB2_ipdtmpl_index and GRIB2_ipdtmpl_val settings are looking for. In both instances, the variables are looking at each individual record's information, rather than the file as a whole (i.e. where the records came from in the file), for which one you'd like to process. 

GRIB2_ipdtmpl_index looks at each index in the GRIB records that are returned from your variable search (in this case, two records for variable TCDC at level L0). The idea is to find an index where the two records differ. In this case, index 8 (which is fairly common for GRIB2 files) shows the first difference of values between the two returned records:

> ipdtmpl[15] = 6, 1, 2, 0, 134, 0, 0, 1, 7, 
> ipdtmpl[29] = 6, 1, 2, 0, 134, 0, 0, 1, 6,

GRIB2_ipdtmpl_val is your way of indicating what value should be there. You've indicated you want the first returned record; so the value at index 8 should be "7". Alternatively, if you wanted the second record you'd set GRIB2_ipdtmpl_val to "6".

So your options should be:

`FCST_VAR8_OPTIONS = GRIB_lvl_typ = 200; GRIB2_ipdtmpl_index = [ 8 ]; GRIB2_ipdtmpl_val = [ 7 ];`

I hope this helps with the issue! If it does, please select this as an answer and I'll close the discussion. If you're still running into issues with this particular problem you are always welcome to continue the discussion in this thread and we'll try to help you through it. And as always, if you run into any problems that are unrelated to this particular issue, we encourage you to mark this question as answered and open a new discussion question.

Thanks,

-John O.

**Reply by georgemccabe** [reply]
Hi @michelleharrold, did @j-opatz's answer resolve your issue? If so, I'd like to mark his response as the answer and lock this discussion to prevent future comments.

**Comment by michelleharrold** [comment, marked as answer]
Hi @j-opatz and @georgemccabe -- apologies for the slow response. @j-opatz, thanks for the in-depth response; it was super helpful in understanding the a little bit more about the intricacies of grib records. I did try your suggestion:

```
FCST_VAR8_NAME = TCDC
FCST_VAR8_LEVELS = L0
FCST_VAR8_OPTIONS = GRIB_lvl_typ = 200; GRIB2_ipdtmpl_index = [ 8 ]; GRIB2_ipdtmpl_val = [ 7 ];
```

But got this warning and yielded no output:

```
DEBUG 2: Reading data for TCDC/L0.
WARNING:
WARNING: process_fcst_climo_files() -> no fields matching TCDC/L0 found in file: /scratch2/BMC/det/harrold/ufs-srw_addVars/expt_dirs/test1_gsd_sar_25km_ens/2019061500/mem02/postprd/rrfs.t00z.prslevf022.tm00.grib2
WARNING:
```
Did I do something silly when updating the syntax?

Thanks!
Michelle

**Reply by j-opatz** [reply]
Hi Michelle,

Sorry that suggestion didn't do the trick. I think this may be a syntax issue, as you mentioned. Have you tried the ipdtmpl_index and ipdtmpl_val without the brackets (e.g. the [ ])? When running pb2nc directly on the files, your command should look something like 

`regrid_data_plane rrfs.t00z.prslevf022.tm00.grib2 regridded_file.nc -field 'name="TCDC"; level="L0"; GRIB2_ipdtmpl_index=8; GRIB2_ipdtmpl_val=7;'`

Can you confirm that a direct call to pb2nc works in this fashion? I'm not as familiar with the METplus wrapper version, but since those variables (ipdtmpl) are integers rather than lists in the MET tool, they probably shouldn't have brackets around them.

The GRIB_lvl_typ may also be unnecessary, but I haven't played around with the particular files you're using to know if that's true. If the above command throws an error just add the GRIB_lvl_typ as

`regrid_data_plane rrfs.t00z.prslevf022.tm00.grib2 regridded_file.nc -field 'name="TCDC"; level="L0"; GRIB_lvl_typ=200; GRIB2_ipdtmpl_index=8; GRIB2_ipdtmpl_val=7;'`

If either of the above commands are successful, it's a good bet that the brackets are what's messing things up.

**Reply by michelleharrold** [reply]
Hi John,

Okay, I think I was able to figure out what was going on. I used plot_data_plane and got the syntax correct (for my test file):

`plot_data_plane rrfs.t00z.prslevf022.tm00.grib2 tcdc.ps 'name="TCDC"; level="L0"; GRIB_lvl_typ=200; GRIB2_ipdtmpl_index=8; GRIB2_ipdtmpl_val=22;'`

Note: I could also use brackets around the `GRIB2_ipdtmpl_index` and `GRIB2_ipdtmpl_val` and was able to produce a plots -- so for this example, the brackets don't seem to be an issue.

The `GRIB2_ipdtmpl_val` is my issue -- this value changes depending on the forecast hour; for the example here, I am looking at forecast hour 22, so `GRIB2_ipdtmpl_val=22`. 

I was able to modify my METplus conf file to dynamically change `GRIB2_ipdtmpl_val` based on my forecast hour.

I think I am done bothering you all - thanks!!
Michelle
